woocommerce/woocommerce
# Validate variation attributes in WC_Cart::add_to_cart()

### All Submissions:






### Changes proposed in this Pull Request:



Currently there's only validation of variation attributes inside `WC_Form_Handler::add_to_cart_handler_variable()` by checking `$_REQUEST`, but there's no validation if you are using AJAX or using `WC()->cart->add_to_cart()` directly, allowing inconsistent behavior while adding variable products to the cart.

I hope with those changes fix the issues reported in #25133.

This PR fixes also a small bug when using:

```
?add-to-cart=<variation_id>&attribute_color=&attribute_size=
?add-to-cart=<parent_product_id>&variation_id=<variation_id>&attribute_color=&attribute_size=
```

It allows products with `any` attributes to get included into the cart without asking for an attribute.


Closes #25133.

### How to test the changes in this Pull Request:

1. Download and import a sample variable product to test: [issue-25133.zip](https://github.com/woocommerce/woocommerce/files/4974566/issue-25133.zip)
2. Try to add one of the variations to the cart using `?add-to-cart=<variation_id>&attribute_size=`
3. Note that the variation got added to the cart with missing attributes.
4. Now try with `?add-to-cart=<parent_product_id>&variation_id=<variation_id>&attribute_size=`
5. Note the missing Size attribute again.
6. Now apply the code of this PR and try again.
7. You should see a message that the "Size" attribute is missing.
8. It's possible to also test using `WC()->cart->add_to_cart( $product_id, $quantity, $variation_id, $attributes_list )`.

### Other information:




### Changelog entry

> Fix - Fixed validation of variation attributes while adding products to the cart.
